README.md
# hbase-jruby

*hbase-jruby* is a simple JRuby binding for HBase.

*hbase-jruby* provides the followings:
- Easy, Ruby-esque interface for the fundamental HBase operations
- ActiveRecord-like method chaining for data retrieval

## Installation

    gem install hbase-jruby

### Using hbase-jruby in HBase shell

You can use this gem in HBase shell without external JRuby installation.

First, clone this repository,

```sh
git clone --depth 1 https://github.com/junegunn/hbase-jruby.git
```

then start up the shell (`hbase shell`) and type in the following line:

```ruby
$LOAD_PATH << 'hbase-jruby/lib'; require 'hbase-jruby'
```

Now, you're all set.

```ruby
# Start using it!
hbase = HBase.new

hbase.list

hbase[:my_table].create! :f
hbase[:my_table].put 100, 'f:a' => 1, 'f:b' => 'two', 'f:c' => 3.14
hbase[:my_table].get(100).double('f:c') # Returns 3.14
```

## A quick example

```ruby
require 'hbase-jruby'

# HBase client dependencies
$CLASSPATH << 'hbase-client-dep-1.0.jar'

# Connect to HBase
hbase = HBase.new 'localhost'

# Table object
table = hbase[:test_table]
table.drop! if table.exists?
table.create! :cf1 => {}, :cf2 => {}

# PUT
table.put 'rowkey1' => { 'cf1:a' => 100, 'cf2:b' => 'Hello' },
          'rowkey2' => { 'cf1:a' => 200, 'cf2:b' => 'world' }

# GET
row = table.get('rowkey1')
number = row.fixnum('cf1:a')
string = row.string('cf1:b')

# SCAN
table.range('rowkey1'..'rowkey9')
     .filter('cf1:a' => 100..200,         # cf1:a between 100 and 200
             'cf1:b' => 'Hello',          # cf1:b = 'Hello'
             'cf2:c' => /world/i,         # cf2:c matches /world/i
             'cf2:d' => ['foo', /^BAR/i]) # cf2:d = 'foo' OR matches /^BAR/i
     .project('cf1:a', 'cf2').
     .each do |row|
  puts row.fixnum('cf1:a')
end

# DELETE
table.delete('rowkey9')
```

## A quick example using schema definition

```ruby
require 'hbase-jruby'

# HBase client dependencies
$CLASSPATH << 'hbase-client-dep-1.0.jar'

# Connect to HBase on localhost
hbase = HBase.new

# Define table schema for easier data access
hbase.schema = {
  # Schema for `book` table
  book: {
    # Columns in cf1 family
    cf1: {
      title:     :string,     # String (UTF-8)
      author:    :string,
      category:  :string,
      year:      :short,      # Short integer (2-byte)
      pages:     :int,        # Integer (4-byte)
      price:     :bigdecimal, # BigDecimal
      height:    :float,      # Single-precision floating-point number (4-byte)
      weight:    :double,     # Double-precision floating-point number (8-byte)
      in_print:  :boolean,    # Boolean (true | false)
      image:     :raw         # Java byte array; no automatic type conversion
      thumbnail: :byte_array  # HBase::ByteArray
    },
    # Columns in cf2 family
    cf2: {
      summary:  :string,
      reviews:  :fixnum,      # Long integer (8-byte)
      stars:    :fixnum,
      /^comment\d+/ => :string
    }
  }
}

# Create book table with two column families
table = hbase[:book]
unless table.exists?
  table.create! cf1: { min_versions: 2 },
                cf2: { bloomfilter: :rowcol, versions: 5 }
end

# PUT
table.put 1,
  title:     'The Golden Bough: A Study of Magic and Religion',
  author:    'Sir James G. Frazer',
  category:  'Occult',
  year:      1890,
  pages:     1006,
  price:     BigDecimal('21.50'),
  weight:    3.0,
  in_print:  true,
  image:     File.open('thumbnail.png', 'rb') { | f          | f.read }.to_java_bytes,
  summary:   'A wide-ranging, comparative study of mythology and religion',
  reviews:   52,
  stars:     226,
  comment1:  'A must-have',
  comment2:  'Rewarding purchase'

# GET (using schema)
book     = table.get(1)
title    = book[:title]
comment2 = book[:comment2]
as_hash  = book.to_h

# GET (not using schema)
title    = book.string('cf1:title')       # cf:cq notation
year     = book.short('cf1:year')
reviews  = book.fixnum('cf2:reviews')
stars    = book.fixnum(['cf2', 'stars'])  # Array notation of [cf, cq]

# SCAN
table.range(0..100)
     .project(:cf1, :reviews, :summary)
     .filter(year:     1880...1900,
             in_print: true,
             category: ['Comics', 'Fiction', /cult/i],
             price:    { lt: BigDecimal('30.00') },
             summary:  /myth/i)
     .each do                                   | book       |

  # Update columns
  table.put book.rowkey, price: book[:price] + BigDecimal('1')

  # Atomic increment
  table.increment book.rowkey, reviews: 1, stars: 5

  # Delete two columns
  table.delete book.rowkey, :comment1, :comment2
end

# Delete row
table.delete 1
```

## Setting up

### Resolving Hadoop/HBase dependency

To be able to access HBase from JRuby, Hadoop/HBase dependency must be
satisfied. This can be done either by setting up CLASSPATH beforehand (e.g.
`CLASSPATH=$(hbase classpath) jruby ...`) or by `require`ing relevant JAR
files after launching JRuby.

You might want to check out pre-built uberjars for various versions of HBase
client in [hbase-client-dep releases page][client].

```ruby
require 'hbase-jruby'
$CLASSPATH << 'hbase-client-dep-1.0.jar'

hbase = HBase.new
```

[client]: https://github.com/junegunn/hbase-client-dep/releases

### Log4j logs from HBase

You can suppress (or customize) log messages from HBase.

```ruby
# With an external log4j.properties or log4j.xml file
HBase.log4j = '/your/log4j.properties'
HBase.log4j = '/your/log4j.xml'

# With a Hash
HBase.log4j = { 'log4j.threshold' => 'ERROR' }
```

### Connecting to HBase

```ruby
# HBase on localhost
hbase = HBase.new

# HBase on remote host
hbase = HBase.new 'remote-server.mydomain.net'

# The above is equivalent to the following:
hbase = HBase.new 'hbase.zookeeper.quorum' => 'remote-server.mydomain.net'

# Extra configuration
hbase = HBase.new 'hbase.zookeeper.quorum'       => 'remote-server.mydomain.net',
                  'hbase.client.retries.number'  => 3,
                  'hbase.client.scanner.caching' => 1000,
                  'hbase.rpc.timeout'            => 120000

# Close HBase connection
hbase.close
```

## Accessing data with HBase::Table instance

`HBase#[]` method (or `HBase#table`) returns an `HBase::Table` instance
which represents the table of the given name.

```ruby
table = hbase.table(:test_table)

# Or simply,
table = hbase[:test_table]
```

### Creating a table

```ruby
# Drop table if exists
table.drop! if table.exists?

# Create table with two column families
table.create! cf1: {},
              cf2: { compression: :snappy, bloomfilter: :row }
```

## List of operations

| Operation          | Description                                                                                   |
| ------------------ | --------------------------------------------------------------------------------------------- |
| PUT                | Puts data into the table                                                                      |
| GET                | Retrieves data from the table by one or more rowkeys                                          |
| SCAN               | Scans the table for a given range of rowkeys                                                  |
| DELETE             | Deletes data in the table                                                                     |
| INCREMENT          | Atomically increments one or more columns                                                     |
| APPEND             | Appends values to one or more columns within a single row                                     |
| Checked PUT/DELETE | Atomically checks if the pre-exising data matches the expected value and puts or deletes data |
| MUTATE             | Performs multiple mutations (PUTS and DELETES) atomically on a single row                     |
| Batch execution    | Performs multiple actions (PUT, GET, DELETE, INCREMENT, APPEND) at once                       |

### Defining table schema for easier data access

HBase stores everything as plain Java byte arrays. So it's completely up to
users to encode and decode column values of various types into and from byte
arrays, and that is a quite tedious and error-prone task.

To remedy this situation, `hbase-jruby` implements the concept of table schema.

Using table schema greatly simplifies the way you access data:
- With schema, byte array conversion becomes automatic
- It allows you to omit column family names (e.g. `:title` instead of `"cf1:title"`)

We'll use the following schema throughout the examples.

```ruby
hbase.schema = {
  # Schema for `book` table
  book: {
    # Columns in cf1 family
    cf1: {
      title:    :string,     # String (UTF-8)
      author:   :string,
      category: :string,
      year:     :short,      # Short integer (2-byte)
      pages:    :int,        # Integer (4-byte)
      price:    :bigdecimal, # BigDecimal
      weight:   :double,     # Double-precision floating-point number
      in_print: :boolean,    # Boolean (true | false)
      image:    :raw         # Java byte array; no automatic type conversion
    },
    # Columns in cf2 family
    cf2: {
      summary:  :string,
      reviews:  :fixnum,     # Long integer (8-byte)
      stars:    :fixnum,
      /^comment\d+/ => :string
    }
  }
}
```

Columns that are not defined in the schema can be referenced
using `FAMILY:QUALIFIER` notation or 2-element Array of column family name (as
Symbol) and qualifier, however since there's no type information, they are
returned as Java byte arrays, which have to be decoded manually.

### PUT

```ruby
# Putting a single row
# - Row keys can be of any type, in this case, we use String type
table.put 'rowkey1', title: "Hello World", year: 2013

# Putting multiple rows
table.put 'rowkey1' => { title: 'foo',    year: 2013 },
          'rowkey2' => { title: "bar",    year: 2014 },
          'rowkey3' => { title: 'foobar', year: 2015 }

# Putting values with timestamps
table.put 'rowkey1',
  title: {
    1353143856665 => "Hello world",
    1352978648642 => "Goodbye world"
  },
  year: 2013
```

### GET

```ruby
book = table.get('rowkey1')

# Rowkey
rowkey = row.rowkey         # Rowkey as raw Java byte array
rowkey = row.rowkey :string # Rowkey as String

# Access columns in schema
title  = book[:title]
author = book[:author]
year   = book[:year]

# Convert to simple Hash
hash = book.to_h

# Convert to Hash containing all versions of values indexed by their timestamps
all_hash = table.versions(:all).get('rowkey1').to_H

# Columns not defined in the schema are returned as Java byte arrays
# They need to be decoded manually
extra = HBase::Util.from_bytes(:bigdecimal, book['cf2:extra'])
# or, simply
extra = book.bigdecimal 'cf2:extra'
```

#### Batch-GET

```ruby
# Pass an array of row keys as the parameter
books = table.get(['rowkey1', 'rowkey2', 'rowkey3'])
```

#### `to_h`

`to_h` and `to_H` return the Hash representation of the row.
(The latter returns all values with their timestamp)

If a column is defined in the schema, it is referenced using its quailifier in
Symbol type. If a column is not defined, it is represented as a 2-element Array
of column family in Symbol and column qualifier as ByteArray.
Even so, to make it easier to reference those columns, an extended version of
Hash is returned with which you can also reference them with `FAMILY:QUALIFIER`
notation or `[cf, cq]` array notation.

```ruby
table.put 1000,
  title:      'Hello world', # Known column
  comment100: 'foo',         # Known column
  'cf2:extra' => 'bar',      # Unknown column
  [:cf2, 10]  => 'foobar'    # Unknown column, non-string qualifier

book = table.get 10000
hash = book.to_h
  # {
  #   :title => "Hello world",
  #   [:cf2, HBase::ByteArray<0, 0, 0, 0, 0, 0, 0, 10>] =>
  #       byte[102, 111, 111, 98, 97, 114]@6f28bb44,
  #   :comment100 => "foo",
  #   [:cf2, HBase::ByteArray<101, 120, 116, 114, 97>] =>
  #       byte[98, 97, 114]@77190cfc}
  # }

hash['cf2:extra']
  # byte[98, 97, 114]@77190cfc

hash[%w[cf2 extra]]
  # byte[98, 97, 114]@77190cfc

hash[[:cf2, HBase::ByteArray['extra']]]
  # byte[98, 97, 114]@77190cfc

hash['cf2:extra'].to_s
  # 'bar'

# Columns with non-string qualifiers must be referenced using 2-element Array notation
hash['cf2:10']
  # nil
hash[[:cf2, 10]]
  # byte[102, 111, 111, 98, 97, 114]@6f28bb44

hash_with_versions = table.versions(:all).get(10000).to_H
  # {
  #   :title => {1369019227766 => "Hello world"},
  #   [:cf2, HBase::ByteArray<0, 0, 0, 0, 0, 0, 0, 10>] =>
  #       {1369019227766 => byte[102, 111, 111, 98, 97, 114]@6f28bb44},
  #   :comment100 => {1369019227766 => "foo"},
  #   [:cf2, HBase::ByteArray<101, 120, 116, 114, 97>]  =>
  #       {1369019227766 => byte[98, 97, 114]@77190cfc}}
  # }
```

#### Intra-row scan

Intra-row scan can be done using `each` method which yields `HBase::Cell` instances.

```ruby
# Intra-row scan (all versions)
row.each do |cell|
  family    = cell.family
  qualifier = cell.qualifier :string  # Column qualifier as String
  timestamp = cell.timestamp
  value     = cell.value
end

# Array of HBase::Cells
cells = row.to_a
```

### DELETE

```ruby
# Delete a row
table.delete('rowkey1')

# Delete all columns in the specified column family
table.delete('rowkey1', :cf1)

# Delete a column
table.delete('rowkey1', :author)

# Delete multiple columns
table.delete('rowkey1', :author, :title, :image)

# Delete a column with empty qualifier.
# (!= deleing the entire columns in the family. See the trailing colon.)
table.delete('rowkey1', 'cf1:')

# Delete a version of a column
table.delete('rowkey1', :author, 1352978648642)

# Delete multiple versions of a column
table.delete('rowkey1', :author, 1352978648642, 1352978649642)

# Delete multiple versions of multiple columns
# - Two versions of :author
# - One version of :title
# - All versions of :image
table.delete('rowkey1', :author, 1352978648642, 1352978649642, :title, 1352978649642, :image)

# Batch delete; combination of aforementioned arguments each given as an Array
table.delete(['rowkey1'], ['rowkey2'], ['rowkey3', :author, 1352978648642, 135297864964])
```

However, the last syntax seems a bit unwieldy when you just wish to delete a few rows.
In that case, use simpler `delete_row` method.

```ruby
table.delete_row 'rowkey1'

table.delete_row 'rowkey1', 'rowkey2', 'rowkey3'
```

### INCREMENT: Atomic increment of column values

```ruby
# Atomically increase cf2:reviews by one
inc = table.increment('rowkey1', reviews: 1)
puts inc[:reviews]

# Atomically increase two columns by one and five respectively
inc = table.increment('rowkey1', reviews: 1, stars: 5)
puts inc[:stars]
```

### APPEND

```ruby
ret = table.append 'rowkey1', title: ' (limited edition)', summary: ' ...'
puts ret[:title]   # Updated title
```

### Checked PUT and DELETE

```ruby
table.check(:rowkey, in_print: false)
     .put(in_print: true, price: BigDecimal('10.0'))

table.check(:rowkey, in_print: false)
     .delete(:price, :image)
       # Takes the same parameters as those of HBase::Table#delete
       # except for the first rowkey
       #   https://github.com/junegunn/hbase-jruby#delete
```

### MUTATE: Atomic mutations on a single row (PUTs and DELETEs)

```ruby
# Currently Put and Delete are supported
# - Refer to mutateRow method of org.apache.hadoop.hbase.client.HTable
table.mutate(rowkey) do |m|
  m.put comment3: 'Nice', comment4: 'Great'
  m.delete :comment1, :comment2
end
```

### Batch execution

*Disclaimer*: The ordering of execution of the actions is not defined.
Refer to the documentation of batch method of [HTable class](http://hbase.apache.org/apidocs/org/apache/hadoop/hbase/client/HTable.html).

```ruby
ret = table.batch do |b|
  b.put rowkey1, 'cf1:a' => 100, 'cf1:b' => 'hello'
  b.get rowkey2
  b.append rowkey3, 'cf1:b' => 'world'
  b.delete rowkey3, 'cf2', 'cf3:z'
  b.increment rowkey3, 'cf1:a' => 200, 'cf1:c' => 300
end
```

`batch` method returns an Array of Hashes which contains the results of the
actions in the order they are specified in the block. Each Hash has `:type` entry
(:get, :put, :append, etc.) and `:result` entry. If the type of an action is
:put or :delete, the `:result` will be given as a boolean. If it's an
:increment or :append, a plain Hash will be returned as the `:result`, just like
in [increment](https://github.com/junegunn/hbase-jruby#increment-atomic-increment-of-column-values)
and [append](https://github.com/junegunn/hbase-jruby#append) methods.
For :get action, `HBase::Row` instance will be returned or nil if not found.

If one or more actions has failed, `HBase::BatchException` will be raised.
Although you don't get to receive the return value from batch method,
you can still access the partial results using `results` method of
`HBase::BatchException`.

```ruby
results =
  begin
    table.batch do |b|
      # ...
    end
  rescue HBase::BatchException => e
    e.results
  end
```

### SCAN

`HBase::Table` itself is an enumerable object.

```ruby
# Full scan
table.each do |row|
  p row.to_h
end

# Returns Enumerator when block is not given
table.each.with_index.each_slice(10).to_a
```

## Scoped access

You can control how you retrieve data by chaining
the following methods of `HBase::Table` (or `HBase::Scoped`).

| Method           | Description                                                     |
|------------------|-----------------------------------------------------------------|
| `range`          | Specifies the rowkey range of scan                              |
| `project`        | To retrieve only a subset of columns                            |
| `filter`         | Filtering conditions of scan                                    |
| `while`          | Allows early termination of scan (server-side)                  |
| `at`             | Only retrieve data with the specified timestamp                 |
| `time_range`     | Only retrieve data within the specified time range              |
| `limit`          | Limits the number of rows                                       |
| `versions`       | Limits the number of versions of each column                    |
| `caching`        | Sets the number of rows for caching during scan                 |
| `batch`          | Limits the maximum number of values returned for each iteration |
| `with_java_scan` | *(ADVANCED)* Access Java Scan object in the given block         |
| `with_java_get`  | *(ADVANCED)* Access Java Get object in the given block          |

Each invocation to these methods returns an `HBase::Scoped` instance with which
you can retrieve data with the following methods.

| Method      | Description                                                             |
|-------------|-------------------------------------------------------------------------|
| `get`       | Fetches rows by the given rowkeys                                       |
| `each`      | Scans the scope of the table (`HBase::Scoped` instance is `Enumerable`) |
| `count`     | Efficiently counts the number of rows in the scope                      |
| `aggregate` | Performs aggregation using Coprocessor (To be described shortly)        |

### Example of scoped access

```ruby

CHANGELOG.md
Changelog
=========

0.7.1
-----

- Support setting one timestamp for all columns (#41)

0.7.0
-----

- Distinguish float and double
    - `Cell#float` / `Cell#double`
    - `Row#float` / `Row#double`
    - `Row#floats` / `Row#doubles`
    - `Util.to_bytes 3.14` (double by default)
    - `Util.to_bytes :float => 3.14`
    - `Util.to_bytes :double => 3.14`
    - `Util.from_bytes :float, bytes'`
- Removed automatic disable/enable to support online alter
- Fixed `HBase::Table#alter_family[!]` not to drop previous non-default
  settings
- Performance improvement with column key interpretation cache
- Dependency resolution is now deprecated
- HBase 1.0 client compatibility

0.6.4
-----

### Performance improvement (experimental)

`HBase::Table` instance can be set up to cache the interpretations of the
column keys using thread-locals which can lead to 3-times faster Put
generation in tight loops.

```ruby
table = hbase[:my_table, cache: true]
# ...
table.close
```

However, the option is off by default because of the following issues:

1. You should not use it with the tables with unlimited number of columns.
2. Caching does not keep track of the updates of the schema. If you change the
   schema of an `HBase` object after you created an `HBase::Table` object with
   the cache turned on from it, the object may hold stale information on the
   types of the columns. `HBase::Table#close` method will clean up the cache.

0.6.3
-----

- Refactoring
    - Changed `HBase::Row#to_h` call to return a Hash extended with
      `HBase::Row::HashExtension` module which makes it possible to extend the
      returned Hash by monkey-patching the module
    - `HBase::Row::HashExtension` overrides `#[]` method to allow looking up data
      with both Array and String notation of column keys: `[CF, CQ]` or `CF:CQ`
- Hash for describing column and table options can now contain XML entries
  using additional `:config` key.

```ruby
hbase[:my_table].create! :d,
    # HTableDescriptor#setValue
   'MAX_FILESIZE' => 1 << 33,
   :config => {
     # HTableDescriptor#setConfiguration
     'hbase.hstore.compaction.max.size' => 1 << 30
   }
```

0.6.2
-----

- Removed automatic cleanup of thread-local HTable instances
    - Fixes issue #35
    - Close of individual HTable instance turned out to be unnecessary
        - `flushCommits` not required as `hbase-jruby` runs in auto-flush mode
        - No need to close shared connection or its thread pool

0.6.1
-----

- Fixed error on JRuby 1.7.13 (#34)

0.6.0
-----

0.6.0 includes a few backward-incompatible changes.

- Default number of versions for a scan or get is now changed to 1 to match Java API
    - Make sure to set versions to `:all` if you're calling `to_H` or plural datatype methods
        - `scoped.versions(:all).to_H`
        - `scoped.versions(:all).strings('d:title')`
- Removed `HBase::Table::BatchAction#mutate` method
    - RowMutation is not allowed in `HBase::Table#batch`
        - https://issues.apache.org/jira/browse/HBASE-11421
- Added CDH5.1 dependency profile

0.5.1
-----
- Fixed `Table#aggregate` for HBase 0.98 or above

0.5.0
-----
- Added CDH5.0 (HBase 0.96.1.1) and CDH4.6 dependency profiles
    - If you share HBase instance between threads, you may need to set
      `hbase.hconnection.threads.core` due to
      [a bug](https://issues.apache.org/jira/browse/HBASE-10449) in CDH5.5.0.
- Use HConnection instead of deprecated HTablePool when possible

0.4.6
-----
- [#29 Fix possible HTable leaks](https://github.com/junegunn/hbase-jruby/issues/29)
- [#30 Make `HBase.log4j=` callable before dependencies are met](https://github.com/junegunn/hbase-jruby/issues/30)

0.4.5
-----
- Fixed HBase 0.96 compatibily issues and tested on HBase 0.96 and 0.98
- Added `:split_policy` table property
- `Table#properties` and `Table#families` now include previously unknown
  properties as String-String pairs
- Added `Row#byte_array` which returns an instance of `HBase::ByteArray`
    - Equivalent to `HBase::ByteArray[row.raw(col)]`


0.4.4
-----
- Fixed `HBase::Table#raw_families/raw_properties` on HBase shell

0.4.3
-----
- Fixed `HBase::Table#add_coprocessor[!]` and `HBase::Table#remove_coprocessor[!]`

0.4.2
-----
- Fixed bug when using schema with non-String/Symbol qualifier
- Updated dependency profiles
  - Added CDH4.5, CDH4.4 (same as CDH4.3)
  - 0.95, 0.96 (experimental, currently not working)
- Improved compatibility with HBase shell which is based on JRuby 1.6.5

0.4.1
-----
- Fixed .META. scanning with range prefix ([#26](https://github.com/junegunn/hbase-jruby/issues/26))
- Added `ByteArray#as` as a synonym for `ByteArray#decode`

0.4.0
-----
- Added support for append operation: `HBase::Table#append`
- Added support for atomic mutations on a single row: `HBase::Table#mutate`
- Added support for batch operations: `HBase::Table#batch`
  - This method does not take an argument and requires a block
  - Don't be confused with the shortcut method to `HBase::Scoped#batch(batch_size)`
- Changed `HBase::Table#increment` to return the updated values as a Hash
- Fixed HBase.resolve_dependency!(:local) on CDH distribution
- Empty-qualifier must be given as 'cf:', and not 'cf'
- Added `HBase::Row#empty?` method
- Added `HBase::ByteArray#to_s` method

0.3.5
-----
- Improved `Scoped#count` method
  - KeyOnlyFilter turned out to be compatible with SingleColumnValueFilter
  - Now takes an optional Hash: `scoped.count(cache_blocks: false, caching: 100)`
  - Changed not to disable server-side block caching by default
- Supports `Scoped#limit` even when `Scan.setMaxResultSize` is not implemented
  - `Scoped#limit(nil)` will remove the previously set value

0.3.4
-----
- Cleanup all thread-local (fiber-local) HTable references when connection is closed
- Added `HBase#reset_table_pool` method for recreating HTable pool

0.3.3
-----

0.3.3 changes the way null values are handled, and introduces interface for CAS operations.

lib/hbase-jruby/table.rb
require 'bigdecimal'
require 'thread'

class HBase
# @!attribute [r] name
#   @return [String] The name of the table
# @!attribute [r] config
#   @return [org.apache.hadoop.conf.Configuration]
class Table
  attr_reader :name
  attr_reader :config

  include Enumerable
  include Admin
  include Scoped::Aggregation::Admin
  include HBase::Util

  # (INTERNAL) Returns the underlying org.apache.hadoop.hbase.client.HTable object (local to current thread)
  # @return [org.apache.hadoop.hbase.client.PooledHTable]
  def htable
    check_closed

    # [:hbase_jruby][HBase connection][Table name]
    thread_local(@hbase, @name_sym)[:htable] ||= @hbase.send(:get_htable, @name)
  end

  # Clean up thread-locals
  # Generally this is not required unless you use unlimited number of threads
  # @return [nil]
  def close
    hash = thread_local(@hbase).delete(@name_sym)
    hash[:htable].close if hash && hash[:htable]
    nil
  end

  # Returns whether if the connection is closed
  # @return [Boolean]
  def closed?
    @hbase.closed?
  end

  [:get, :count, :aggregate,
   :range, :project, :filter, :while,
   :limit, :versions, :caching, # :batch,
   :time_range, :at
  ].each do |method|
    define_method(method) do |*args|
      self.scoped.send(method, *args)
    end
  end

  def with_java_scan &block
    self.scoped.with_java_scan(&block)
  end

  def with_java_get &block
    self.scoped.with_java_get(&block)
  end

  # Performs PUT operations
  # @overload put(rowkey, data)
  #   Put operation on a rowkey
  #   @param [Object] rowkey Rowkey
  #   @param [Hash] data Data to put
  #   @return [Fixnum] Number of puts succeeded
  # @overload put(data)
  #   Put operation on multiple rowkeys
  #   @param [Hash<Hash>] data Data to put indexed by rowkeys
  #   @return [Fixnum] Number of puts succeeded
  def put *args
    case args.length
    when 1
      puts = args.first.map { |rowkey, props| @mutation.put rowkey, props }
      htable.put puts
      puts.length
    when 2, 3
      htable.put @mutation.put(*args)
      1
    else
      raise ArgumentError, 'invalid number of arguments'
    end
  end

  # Deletes data
  # @overload delete(rowkey)
  #   Deletes a row with the given rowkey
  #   @param [Object] rowkey
  #   @return [nil]
  #   @example
  #     table.delete('a000')
  # @overload delete(rowkey, *extra)
  #   Deletes columns in the row.
  #   @param [Object] rowkey
  #   @param [*Object] extra [Family|Qualifier [Timestamp ...]] ...
  #   @return [nil]
  #   @example
  #     # A column (with schema)
  #     table.delete('a000', :title)
  #     # Two columns (with schema)
  #     table.delete('a000', :title, :author)
  #     # A column (without schema)
  #     table.delete('a000', 'cf1:col1')
  #     # Columns in cf1 family
  #     table.delete('a000', 'cf1')
  #     # A version
  #     table.delete('a000', :author, 1352978648642)
  #     # Two versions
  #     table.delete('a000', :author, 1352978648642, 1352978647642)
  #     # Combination of columns and versions
  #     table.delete('a000', :author, 1352978648642, 1352978647642,
  #                          :title,
  #                          :image, 1352978648642)
  # @overload delete(*delete_specs)
  #   Batch deletion
  #   @param [*Array] delete_specs
  #   @return [nil]
  #   @example
  #     table.delete(
  #       ['a000', 'cf1:col1', 1352978648642],
  #       ['a001', 'cf1:col1'],
  #       ['a002', 'cf1'],
  #       ['a003'])
  def delete *args
    specs = args.first.is_a?(Array) ? args : [args]

    list = specs.map { |spec| spec.empty? ? nil : @mutation.delete(*spec) }.compact
    if list.length == 1
      htable.delete list.first
    else
      htable.delete list
    end
  end

  # Delete rows.
  # @param [*Object] rowkeys List of rowkeys of rows to delete
  # @return [nil]
  def delete_row *rowkeys
    list = rowkeys.map { |rk| Delete.new(Util.to_bytes rk) }
    if list.length == 1
      htable.delete list.first
    else
      htable.delete list
    end
  end

  # Atomically increase numeric values
  # @overload increment(rowkey, column, by)
  #   Atomically increase column value by the specified amount
  #   @param [Object] rowkey Rowkey
  #   @param [String, Array] column Column expression in String "FAMILY:QUALIFIER", or in Array [FAMILY, QUALIFIER]
  #   @param [Fixnum] by Increment amount
  #   @return [Hash]
  #   @example
  #     table.increment('a000', 'cf1:col1', 1)
  # @overload increment(rowkey, column_by_hash)
  #   Atomically increase values of multiple columns
  #   @param [Object] rowkey Rowkey
  #   @param [Hash] column_by_hash Column expression to increment amount pairs
  #   @return [Hash]
  #   @example
  #     table.increment('a000', 'cf1:col1' => 1, 'cf1:col2' => 2)
  # @overload increment(inc_spec)
  #   Increase values of multiple columns from multiple rows.
  #   @param [Hash] inc_spec { rowkey => { col => by } }
  #   @return [Hash]
  #   @example
  #     table.increment 'a000' => { 'cf1:col1' => 1, 'cf1:col2' => 2 },
  #                     'a001' => { 'cf1:col1' => 3, 'cf1:col2' => 4 }
  def increment rowkey, *args
    if args.empty? && rowkey.is_a?(Hash)
      INSENSITIVE_ROW_HASH.clone.tap { |result|
        rowkey.each do |key, spec|
          result[Util.java_bytes?(key) ? ByteArray[key] : key] = increment(key, spec)
        end
      }
    else
      incr = @mutation.increment(rowkey, *args)
      Row.send(:new, self, htable.increment(incr)).to_h.tap { |h|
        h.each do |k, v|
          h[k] = Util.from_bytes :fixnum, v unless v.is_a?(Fixnum)
        end
      }
    end
  end

  # Appends values to one or more columns within a single row.
  # @param [Object] rowkey Rowkey
  # @param [Hash] spec Hash (column to value)
  # @return [Hash] Updated values
  # @example
  #   table.put :rowkey, col1: 'hello', col2: 'foo'
  #   table.append :rowkey, col1: ' world', col2: 'bar'
  #     # { col1: 'hello world', col2: 'foobar' }
  def append rowkey, spec
    result = htable.append @mutation.append(rowkey, spec)
    Row.send(:new, self, result).to_h if result # (maybe null)
  end

  # Performs multiple mutations atomically on a single row.
  # Currently Put and Delete are supported.
  # The mutations are performed in the order in which they were specified.
  # @param [Object] rowkey Rowkey
  # @yield [HBase::Table::Mutation::Mutator]
  # @return [nil]
  # @example
  #   table.mutate do |m|
  #     m.put a: 100, b: 'hello'
  #     m.delete :c, :d
  #     m.put e: 3.14
  #   end
  def mutate(rowkey, &block)
    ms = @mutation.mutate(rowkey, &block)
    htable.mutateRow ms if ms
  end

  # Scan through the table
  # @yield [row] Yields each row in the scope
  # @yieldparam [HBase::Row] row
  # @return [Enumerator]
  def each &block
    scoped.each(&block)
  end

  # Returns HBase::Scoped object for this table
  # @return [HBase::Scoped]
  def scoped
    Scoped.send(:new, self, @hbase.config.get('hbase.client.scanner.caching').to_i)
  end

  # Returns CheckedOperation instance for check-and-put and check-and-delete
  # @param [Object] rowkey
  # @param [Hash] cond
  # @return [HBase::Table::CheckedOperation]
  def check rowkey, cond
    raise ArgumentError, 'invalid check condition' unless cond.length == 1
    col, val = cond.first

    cf, cq, type = lookup_and_parse(col, true)

    # If the passed value is null, the check is for the lack of column
    CheckedOperation.new self, @mutation, Util.to_bytes(rowkey),
      cf, cq,
      (val.nil? ? nil : Util.to_typed_bytes(type, val))
  end

  # Method that does a batch call on Deletes, Gets, Puts, Increments, Appends
  # and RowMutations. The ordering of execution of the actions is not defined.
  # An Array of Hashes are returned as the results of each operation. For
  # Delete, Put, and RowMutation, :result entry of the returned Hash is
  # Boolean. For Increment and Append, it will be plain Hashes, and for Get,
  # HBase::Rows will be returned.
  # When an error has occurred, you can still access the partial results using
  # `results` method of the thrown BatchException instance.
  # @yield [HBase::Table::BatchAction]
  # @return [Array<Hash>]
  # @raise [HBase::BatchException]
  def batch arg = nil
    if arg
      # Backward compatibility
      return scoped.batch(arg)
    else
      raise ArgumentError, "Block not given" unless block_given?
    end
    b = BatchAction.send(:new, self, @mutation)
    yield b
    results = Array.new(b.actions.length).to_java
    process = lambda do
      results.each_with_index do |r, idx|
        action = b.actions[idx]
        type = action[:type]
        case type
        when :get
          action[:result] = (r.nil? || r.empty?) ? nil : Row.send(:new, self, r)
        when :append
          action[:result] = r && Row.send(:new, self, r).to_h
        when :increment
          action[:result] = r &&
            Row.send(:new, self, r).to_h.tap { |h|
              h.each do |k, v|
                h[k] = Util.from_bytes :fixnum, v unless v.is_a?(Fixnum)
              end
            }
        else
          case r
          when java.lang.Exception
            action[:result] = false
            action[:exception] = r
          when nil
            action[:result] = false
          else
            action[:result] = true
          end
        end
      end
      b.actions
    end

    begin
      htable.batch b.actions.map { |a| a[:action] }, results
      process.call
    rescue Exception => e
      raise HBase::BatchException.new(e, process.call)
    end
  end

  # @private
  def lookup_schema col
    @hbase.schema.lookup @name_sym, col
  end

  # Schema lookup without column-key caching
  # @private
  def lookup_and_parse col, expect_cq
    @hbase.schema.lookup_and_parse @name_sym, col, expect_cq
  end

  # @private
  module ThreadLocalCache
    # Schema lookup with column-key caching
    def lookup_and_parse col, expect_cq
      cache = thread_local(@hbase, @name_sym, :columns)

      # Column-key cache invalidation
      schema = @hbase.schema
      if schema.to_h.object_id != @schema_h.object_id
        @schema_h = schema.to_h
        cache.clear
      end

      unless parsed = cache[col]
        parsed = schema.lookup_and_parse(@name_sym, col, expect_cq)
        cache[col] = parsed if cache.length < @cache_size
      end
      parsed
    end
  end

private
  def initialize hbase, config, name, cache_size
    @hbase    = hbase
    @schema_h = hbase.schema.to_h
    @config   = config
    @name     = name.to_s
    @name_sym = name.to_sym
    @mutation = Mutation.new(self)
    @cache_size = cache_size

    extend ThreadLocalCache if cache_size > 0
  end

  def check_closed
    raise RuntimeError, "HBase connection is already closed" if @hbase.closed?
  end

  INSENSITIVE_ROW_HASH = {}.tap { |h|
    h.instance_eval do
      def [] key
        if Util.java_bytes?(key)
          key = ByteArray[key]
        end
        super key
      end
    end
  }
end#Table
end#HBase

